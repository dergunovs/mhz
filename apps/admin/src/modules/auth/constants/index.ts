export const API_POST_LOGIN = 'auth/login';
export const API_GET_CHECK_AUTH = 'auth/check';

export const URL_LOGIN = '/';
