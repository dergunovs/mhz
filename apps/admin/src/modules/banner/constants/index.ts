export const URL_BANNER = '/banners';
export const URL_BANNER_CREATE = `${URL_BANNER}/create`;
export const URL_BANNER_EDIT = `${URL_BANNER}/edit`;
