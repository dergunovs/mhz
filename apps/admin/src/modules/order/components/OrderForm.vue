<template>
  <div>{{ props.order }}</div>
</template>

<script setup lang="ts">
import { IOrder } from 'mhz-types';

interface IProps {
  order: IOrder;
}

const props = defineProps<IProps>();
</script>
