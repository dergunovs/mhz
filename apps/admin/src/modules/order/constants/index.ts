export const URL_ORDER = '/orders';
