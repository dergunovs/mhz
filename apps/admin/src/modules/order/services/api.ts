import { ComputedRef, Ref } from 'vue';

import { api, IPageQuery, convertParams } from 'mhz-helpers';
import { API_ORDER, IBaseReply, IOrder, TOrderStatus } from 'mhz-contracts';

export async function getOrdersApi(query: Ref<IPageQuery | number>) {
  const params = convertParams(query);

  const { data } = await api.get<{ data: IOrder[]; total: number }>(API_ORDER, { params });

  return data;
}

export async function getOrderApi(id?: ComputedRef<string | string[]>) {
  if (!id?.value) return null;

  const { data } = await api.get<{ data: IOrder }>(`${API_ORDER}/${id.value}`);

  return data.data;
}

export async function updateOrderApi(order: { status: TOrderStatus; id?: string }) {
  if (!order.id) return null;

  const { data } = await api.patch<IBaseReply>(`${API_ORDER}/${order.id}`, { status: order.status });

  return data;
}

export async function deleteOrderApi(id?: string) {
  if (!id) return null;

  const { data } = await api.delete<IBaseReply>(`${API_ORDER}/${id}`);

  return data;
}
