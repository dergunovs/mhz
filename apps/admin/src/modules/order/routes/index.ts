import { URL_ORDER } from '@/order/constants';

export const orderRoutes = [
  {
    path: URL_ORDER,
    name: 'OrderList',
    component: () => import('@/order/pages/OrderListPage.vue'),
  },
  {
    path: `${URL_ORDER}/:order`,
    name: 'Order',
    component: () => import('@/order/pages/OrderPage.vue'),
  },
];
