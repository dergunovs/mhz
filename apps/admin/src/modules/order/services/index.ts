import { ComputedRef, Ref } from 'vue';

import { IOrder, TOrderStatus } from 'mhz-types';
import { api, useQuery, useMutation, IPageQuery } from 'mhz-helpers';

import { API_ORDER } from '@/order/constants';

export function getOrders(query: Ref<IPageQuery | number>) {
  async function fn(): Promise<{ data: IOrder[]; total: number }> {
    const params =
      typeof query.value === 'number'
        ? { page: query.value }
        : {
            page: query.value.page || 1,
            sort: query.value.sort.value,
            dir: query.value.sort.isAsc === false ? 'desc' : 'asc',
          };

    const { data } = await api.get(API_ORDER, { params });

    return data;
  }

  return useQuery({ queryKey: [API_ORDER, query], queryFn: fn });
}

export function getOrder(id?: ComputedRef<string | string[]>) {
  async function fn(): Promise<IOrder | null> {
    if (!id?.value) return null;

    const { data } = await api.get(`${API_ORDER}/${id.value}`);

    return data;
  }

  return useQuery({ queryKey: [API_ORDER, id], queryFn: fn });
}

export function updateOrder(id: ComputedRef<string | undefined>, options: object) {
  async function fn(status: TOrderStatus) {
    await api.patch(`${API_ORDER}/${id.value}`, { status });
  }

  return useMutation({ mutationKey: [API_ORDER, id], mutationFn: fn, ...options });
}

export function deleteOrder(id: ComputedRef<string | undefined>, options: object) {
  async function fn() {
    await api.delete(`${API_ORDER}/${id.value}`);
  }

  return useMutation({ mutationKey: [API_ORDER], mutationFn: fn, ...options });
}
