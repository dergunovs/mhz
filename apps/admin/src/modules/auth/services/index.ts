import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
