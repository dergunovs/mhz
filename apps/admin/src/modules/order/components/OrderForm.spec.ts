import { ComputedRef, DefineComponent } from 'vue';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount, flushPromises } from '@vue/test-utils';
import { IBaseReply, API_ORDER, TOrderStatus } from 'mhz-contracts';
import { toast } from 'mhz-ui';
import * as helpers from 'mhz-helpers';
import { dataTest } from 'mhz-helpers';

import OrderForm from './OrderForm.vue';

import { wrapperFactory, mockMutationReply, router } from '@/common/test';
import { ORDERS } from '@/order/fixtures';
import { URL_PRODUCT_EDIT } from '@/product/constants';
import { ORDER_CANCELLED, ORDER_COMPLETED, URL_ORDER } from '@/order/constants';
import * as orderServices from '@/order/services';
import { CURRENCY } from '@/common/constants';
import { URL_CUSTOMER } from '@/customer/constants';

const ORDER = ORDERS.data[0];

let onSuccessUpdate: () => void;
const spyMutateUpdate = vi.fn();

vi.spyOn(orderServices, 'updateOrder').mockImplementation(
  (id: ComputedRef<string | undefined>, options: { onSuccess?: () => void }) => {
    if (options.onSuccess) onSuccessUpdate = options.onSuccess;

    return mockMutationReply<IBaseReply, { status: TOrderStatus }>(spyMutateUpdate);
  }
);

let onSuccessDelete: () => void;
const spyMutateDelete = vi.fn();

vi.spyOn(orderServices, 'deleteOrder').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessDelete = options.onSuccess;

  return mockMutationReply<IBaseReply, string | undefined>(spyMutateDelete);
});

const spyRefetchQueries = vi.fn();
const spyRemoveQueries = vi.fn();

vi.spyOn(helpers, 'useQueryClient').mockReturnValue({
  refetchQueries: spyRefetchQueries,
  removeQueries: spyRemoveQueries,
} as unknown as helpers.QueryClient);

const spyRouterPush = vi.spyOn(router, 'push');
const spyRouterGo = vi.spyOn(router, 'go');
const spyToastSuccess = vi.spyOn(toast, 'success');

const QUERY_KEY = { queryKey: [API_ORDER] };

const orderFormCustomer = dataTest('order-form-customer');
const orderFormStatus = dataTest('order-form-status');
const orderFormProducts = dataTest('order-form-products');
const orderFormProductsCount = dataTest('order-form-products-count');
const orderFormProductsTitle = dataTest('order-form-products-title');
const orderFormPrice = dataTest('order-form-price');
const orderFormCreated = dataTest('order-form-created');
const orderFormUpdated = dataTest('order-form-updated');
const orderFormComplete = dataTest('order-form-complete');
const orderFormBack = dataTest('order-form-back');
const orderFormCancel = dataTest('order-form-cancel');
const orderFormCancelConfirm = dataTest('order-form-cancel-confirm');
const orderFormDelete = dataTest('order-form-delete');
const orderFormDeleteConfirm = dataTest('order-form-delete-confirm');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(OrderForm, {
    props: {
      order: ORDER,
    },
  });
});

enableAutoUnmount(afterEach);

describe('OrderForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(OrderForm)).toBeTruthy();
  });

  it('shows order content in fields if props is passed', async () => {
    expect(wrapper.find(orderFormCustomer).text()).toEqual(`${ORDER.customer.firstName} ${ORDER.customer.lastName}`);
    expect(wrapper.find(orderFormCustomer).attributes('to')).toEqual(`${URL_CUSTOMER}/${ORDER.customer._id}`);

    expect(wrapper.find(orderFormStatus).text()).toEqual(ORDER.status);
    expect(wrapper.findAll(orderFormProducts).length).toEqual(ORDER.products?.length);
    expect(wrapper.find(orderFormProductsCount).text()).toEqual(ORDER.products?.[0].count.toString());

    expect(wrapper.find(orderFormProductsTitle).text()).toEqual(ORDER.products?.[0].product.title);
    expect(wrapper.find(orderFormProductsTitle).attributes('to')).toEqual(
      `${URL_PRODUCT_EDIT}/${ORDER.products?.[0].product._id}`
    );

    expect(wrapper.find(orderFormPrice).text()).toEqual(`${ORDER.price} ${CURRENCY}`);
    expect(wrapper.find(orderFormCreated).text()).toEqual(helpers.formatDateTime(ORDER.dateCreated));
    expect(wrapper.find(orderFormUpdated).text()).toEqual(helpers.formatDateTime(ORDER.dateUpdated));
  });

  it('markes order as completed', async () => {
    expect(spyMutateUpdate).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);

    await wrapper.findComponent(orderFormComplete).trigger('click');

    expect(spyMutateUpdate).toBeCalledTimes(1);
    expect(spyMutateUpdate).toBeCalledWith({ status: ORDER_COMPLETED, id: ORDER._id });

    onSuccessUpdate();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);
  });

  it('goes back by back button click', async () => {
    expect(spyRouterGo).toBeCalledTimes(0);

    await wrapper.findComponent(orderFormBack).trigger('click');

    expect(spyRouterGo).toBeCalledTimes(1);
    expect(spyRouterGo).toBeCalledWith(-1);
  });

  it('cancels order', async () => {
    expect(spyMutateUpdate).toBeCalledTimes(0);
    expect(spyRemoveQueries).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    await wrapper.findComponent(orderFormCancel).trigger('click');

    wrapper.findComponent<DefineComponent>(orderFormCancelConfirm).vm.$emit('confirm');

    expect(spyMutateUpdate).toBeCalledTimes(1);
    expect(spyMutateUpdate).toBeCalledWith({ status: ORDER_CANCELLED, id: ORDER._id });

    onSuccessUpdate();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);
  });

  it('deletes order', async () => {
    expect(spyMutateDelete).toBeCalledTimes(0);
    expect(spyRemoveQueries).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    await wrapper.findComponent(orderFormDelete).trigger('click');

    wrapper.findComponent<DefineComponent>(orderFormDeleteConfirm).vm.$emit('confirm');

    expect(spyMutateDelete).toBeCalledTimes(1);
    expect(spyMutateDelete).toBeCalledWith(ORDER._id);

    onSuccessDelete();

    expect(spyRemoveQueries).toBeCalledTimes(1);
    expect(spyRemoveQueries).toBeCalledWith(QUERY_KEY);

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_ORDER);
  });
});
