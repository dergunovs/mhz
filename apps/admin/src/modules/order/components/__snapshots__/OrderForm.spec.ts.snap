// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OrderForm > matches snapshot 1`] = `
"<div class="_container_74860f">
  <div class="_order_74860f">
    <div><b>Customer: </b><a to="/customers/64fdd1910e0355e88d03bdb6" data-test="order-form-customer">Alexandr Dergunov</a></div>
    <div><b>Status: </b><span class="_status_74860f" data-status="new" data-test="order-form-status">new</span></div>
    <div><b>Products: </b>
      <div data-test="order-form-products"><span data-test="order-form-products-count">11</span> x <a to="/products/edit/64dddadbeb94d9f7131a40d0" data-test="order-form-products-title">ASUS ROG Azoth</a></div>
      <div data-test="order-form-products"><span data-test="order-form-products-count">3</span> x <a to="/products/edit/64df860a7ac901ef3e27319c" data-test="order-form-products-title">Thermaltake TOUGHRAM 2x8 GB DDR4 4400</a></div>
    </div>
    <div><b>Price: </b><span data-test="order-form-price">352000 â‚½</span></div>
    <div><b>Created: </b><span data-test="order-form-created">Jan 2, 2024, 1:15 PM</span></div>
    <div><b>Updated: </b><span data-test="order-form-updated">Jan 2, 2024, 1:17 PM</span></div>
  </div>
  <div class="_buttons_74860f">
    <div class="_buttonsInner_74860f"><button isdisabled="true" data-test="order-form-complete"> Mark completed </button><button layout="secondary" data-test="order-form-back">Back</button></div>
    <div class="_buttonsInner_74860f"><button layout="secondary" data-test="order-form-cancel">Cancel order</button><button layout="secondary" data-test="order-form-delete">Delete</button></div>
  </div>
  <div modelvalue="false" isconfirm="" data-test="order-form-cancel-confirm"> Confirm cancel order? </div>
  <div modelvalue="false" isconfirm="" data-test="order-form-delete-confirm">Confirm delete?</div>
</div>"
`;
